kernel/src/expressions/column_name.rs

```diff
@@ -0,0 +1,322 @@
+/// simple (non-nested) column names. For example, the invocation `joined_column_name!("a", "b")` is
+/// equivalent to `ColumnName::join(simple_column_name!("a"), simple_column_name!("b"))`
+///
+/// To avoid accidental misuse, at least one argument must be a string literal. Thus, the following
+/// invocation would fail to compile:
+///
+/// ```fail_compile
+/// # use delta_kernel::expressions::joined_column_name;
+/// let s = "s";
+/// let name = joined_column_name!(s, s);
+/// ```
+///
+/// NOTE: `joined_column_name!("a.b", "c")` produces a path with field names "a.b" and "c". If
+/// either argument is a nested path delimited by periods, it should be split first and the result
+/// passed to `joined_column_name!`, e.g. `joined_column_name!(nested_column_name!("a.b"), "c")`.
+#[macro_export]
+#[doc(hidden)]
+macro_rules! __joined_column_name {
+    ( $left:literal, $right:literal ) => {
+        $crate::__simple_column_name!($left).join(&$crate::__simple_column_name!($right))
+    };
+    ( $left:literal, $right:expr ) => {
+        $crate::__simple_column_name!($left).join(&$right)
+    };
+    ( $left:expr, $right:literal) => {
+        $left.join(&$crate::__simple_column_name!($right))
+    };
+    ( $($other:tt)* ) => {
+        compile_error!("joined_column_name!() requires at least one string literal input")
+    };
+}
+#[doc(inline)]
+pub use __joined_column_name as joined_column_name;
+
+#[macro_export]
+#[doc(hidden)]
+macro_rules! __simple_column_expr {
+    ( $($name:tt)* ) => {
+        $crate::expressions::Expression::from($crate::expressions::simple_column_name!($($name)*))
+    };
+}
+#[doc(inline)]
+pub use __simple_column_expr as simple_column;
+
+#[macro_export]
+#[doc(hidden)]
+macro_rules! __nested_column_expr {
+    ( $($name:tt)* ) => {
+        $crate::expressions::Expression::from($crate::expressions::nested_column_name!($($name)*))
+    };
+}
+#[doc(inline)]
+pub use __nested_column_expr as nested_column;
+
+#[macro_export]
+#[doc(hidden)]
+macro_rules! __joined_column_expr {
+    ( $($name:tt)* ) => {
+        $crate::expressions::Expression::from($crate::expressions::joined_column_name!($($name)*))
+    };
+}
+#[doc(inline)]
+pub use __joined_column_expr as joined_column;
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use delta_kernel_derive::{parse_nested_column_name, parse_simple_column_name};
+
+    #[test]
+    fn test_parse_column_name_macros() {
+        assert_eq!(parse_simple_column_name!("a"), ["a"]);
+
+        assert_eq!(parse_nested_column_name!("a"), ["a"]);
+        assert_eq!(parse_nested_column_name!("a.b"), ["a", "b"]);
+        assert_eq!(parse_nested_column_name!("a.b.c"), ["a", "b", "c"]);
+    }
+
+    #[test]
+    fn test_column_name_macros() {
+        let simple = simple_column_name!("x");
+        let nested = nested_column_name!("x.y");
+
+        // fails to compile (argument is not a literal string, or contains illegal characters)
+        //simple_column_name!("a.b");
```

Comment:
Not sure if we want to test that the failure cases fail to compile.  In case we do, there's a couple crates for that. [compiletest-rs](https://crates.io/crates/compiletest_rs) was used in the rust compiler, but I've heard some folks complain it's hacky. There's also [trybuild](https://crates.io/crates/trybuild/1.0.75) which [serde uses](https://github.com/serde-rs/serde/blob/master/test_suite/tests/compiletest.rs).

Reply:
There are doc tests for this.